pkg/daemon/daemon.go

```diff
@@ -291,6 +315,76 @@ func (dn *Daemon) getHealth() error {
 	return nil
 }
 
+// stateAndConfigs is the "state" node annotation plus parsed machine configs
+// referenced by the currentConfig and desiredConfig annotations.
+type stateAndConfigs struct {
+	state          string
+	currentConfig *mcfgv1.MachineConfig
+	desiredConfig *mcfgv1.MachineConfig
+}
+
+func (dn *Daemon) getStateAndConfigs() (*stateAndConfigs, error) {
+	// TODO: Add a retry loop around these just like we do with getMachineConfig()
+	state, err := getNodeAnnotationExt(dn.kubeClient.CoreV1().Nodes(), dn.name, MachineConfigDaemonStateAnnotationKey, true)
+	if err != nil {
+		return nil, err
+	}
+	currentConfigName, err := getNodeAnnotation(dn.kubeClient.CoreV1().Nodes(), dn.name, CurrentMachineConfigAnnotationKey)
+	if err != nil {
+		return nil, err
+	}
+	desiredConfigName, err := getNodeAnnotation(dn.kubeClient.CoreV1().Nodes(), dn.name, DesiredMachineConfigAnnotationKey)
+	if err != nil {
+		return nil, err
+	}
+	currentConfig, err := getMachineConfig(dn.client.MachineconfigurationV1().MachineConfigs(), currentConfigName)
+	if err != nil {
+		return nil, err
+	}
+	var desiredConfig *mcfgv1.MachineConfig
+	if currentConfigName == desiredConfigName {
+		desiredConfig = currentConfig
+	} else {
+		desiredConfig, err = getMachineConfig(dn.client.MachineconfigurationV1().MachineConfigs(), desiredConfigName)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &stateAndConfigs{
+		currentConfig: currentConfig,
+		desiredConfig: desiredConfig,
+		state: state,
+	}, nil
+}
+
+// isConfigCompletedPending checks whether we've successfully
+// completed a config, and are coming up after a reboot.
+func (dn *Daemon) isConfigCompletedPending(pending string) (bool, error) {
```

Comment:
This function is probably the most important change here; it's a bit subtle but makes sense I think.